### ISSUE TITLE
PyTorch XLA is very slow on Google Colab

### ISSUE THREAD

**Issue by bjourne** [pytorch/xla#2247]
Training a network on Google Colab with PyTorch XLA and with eight TPUs is about 10 times slower than using the GPU and no XLA. This is unexpected. Shouldn't XLA make training much faster?

Here is the training loop of my program which is based on the [PyTorch on Cloud TPUs: MultiCore Training AlexNet on Fashion MNIST](https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/multi-core-alexnet-fashion-mnist.ipynb) notebook:

```python
    for i in range(3):
        start = time()
        loader = ParallelLoader(train_loader, [dev]) \
            .per_device_loader(dev)
        state = model.init_state(batch_size, dev)
        model.train()
        for x, y in loader:
            opt.zero_grad()
            state = [s.detach() for s in state]
            y_hat, state = model(x, state)
            loss = crit(y_hat, y.reshape(-1))
            loss.backward()
            optimizer_step(opt)
        elapsed = time() - start
        master_print('%.2f seconds for %d batches.'
                     % (elapsed, len(train_ds)))
```
Each epoch takes about 140 seconds. An equivalent program running on Google Colab with a GPU takes 20 seconds per epoch. Here is a self-contained gist demonstrating the slowness. https://gist.github.com/bjourne/0cc9ae5729ab78ce0c7765a5c5a207c0 It uses the observations package for the Penn Treebank dataset.

**Comment by btomtom5** [comment]
For easier debugging, I added a metrics report. This was taken at the very last epoch. In this case I ran it for 20 epochs. Slight tangent, but what does XRT<metric_name> mean? I found that XRT is the communication layer between the XLA device and python, but what is the difference between CompileTime and XRTCompile or ExecuteTime and XRTExecute? Thanks! 
```
Metric: CompileTime
  TotalSamples: 4
  Accumulator: 04s589ms590.974us
  ValueRate: 023ms960.897us / second
  Rate: 0.0255932 / second
  Percentiles: 1%=024ms320.066us; 5%=024ms320.066us; 10%=024ms320.066us; 20%=024ms320.066us; 50%=02s720ms247.948us; 80%=02s746ms660.403us; 90%=02s746ms660.403us; 95%=02s746ms660.403us; 99%=02s746ms660.403us
Metric: DeviceLockWait
  TotalSamples: 1122
  Accumulator: 003ms174.641us
  ValueRate: 001.328us / second
  Rate: 0.468417 / second
  Percentiles: 1%=001.701us; 5%=002.097us; 10%=002.230us; 20%=002.349us; 50%=002.587us; 80%=002.903us; 90%=003.121us; 95%=003.431us; 99%=011.410us
Metric: ExecuteTime
  TotalSamples: 1121
  Accumulator: 03m04s242ms643.649us
  ValueRate: 076ms890.245us / second
  Rate: 0.468365 / second
  Percentiles: 1%=054ms856.143us; 5%=054ms447.633us; 10%=055ms898.424us; 20%=063ms551.695us; 50%=145ms486.663us; 80%=241ms377.513us; 90%=294ms087.724us; 95%=340ms904.275us; 99%=463ms020.880us
Metric: InputOutputAliasCount
  TotalSamples: 4
  Accumulator: 51.00
  ValueRate: 0.33 / second
  Rate: 0.0256053 / second
  Percentiles: 1%=2.00; 5%=2.00; 10%=2.00; 20%=2.00; 50%=20.00; 80%=20.00; 90%=20.00; 95%=20.00; 99%=20.00
Metric: IrValueTensorToXlaData
  TotalSamples: 41
  Accumulator: 01s282ms455.891us
  ValueRate: 564.427us / second
  Rate: 0.0180447 / second
  Percentiles: 1%=001ms385.848us; 5%=002ms427.149us; 10%=003ms972.753us; 20%=003ms290.177us; 50%=004ms007.770us; 80%=055ms577.372us; 90%=081ms776.349us; 95%=113ms722.863us; 99%=263ms837.832us
Metric: OutboundData
  TotalSamples: 1439
  Accumulator: 179.70MB
  ValueRate: 74.63KB / second
  Rate: 0.598097 / second
  Percentiles: 1%=8.00B; 5%=8.00B; 10%=8.00B; 20%=8.00B; 50%=8.00B; 80%=160.00KB; 90%=640.00KB; 95%=640.00KB; 99%=640.00KB
Metric: ReleaseDataHandlesTime
  TotalSamples: 11372
  Accumulator: 14s950ms815.598us
  ValueRate: 006ms832.635us / second
  Rate: 5.17792 / second
  Percentiles: 1%=532.212us; 5%=626.633us; 10%=686.950us; 20%=789.423us; 50%=001ms034.373us; 80%=001ms259.645us; 90%=001ms386.729us; 95%=002ms687.614us; 99%=004ms146.956us
Metric: TensorsGraphSize
  TotalSamples: 1122
  Accumulator: 39030977.00
  ValueRate: 16303.02 / second
  Rate: 0.468414 / second
  Percentiles: 1%=10.00; 5%=35322.00; 10%=35322.00; 20%=35322.00; 50%=35322.00; 80%=35322.00; 90%=35322.00; 95%=35322.00; 99%=35322.00
Metric: TransferToServerTime
  TotalSamples: 1439
  Accumulator: 58s794ms674.254us
  ValueRate: 022ms450.591us / second
  Rate: 0.5981 / second
  Percentiles: 1%=002ms723.752us; 5%=002ms020.673us; 10%=002ms265.361us; 20%=003ms958.835us; 50%=019ms916.308us; 80%=058ms724.968us; 90%=105ms078.122us; 95%=150ms419.519us; 99%=240ms860.630us
Metric: TransferToServerTransformTime
  TotalSamples: 1439
  Accumulator: 07s085ms968.392us
  ValueRate: 003ms932.982us / second
  Rate: 0.598097 / second
  Percentiles: 1%=045.284us; 5%=053.907us; 10%=058.274us; 20%=071.031us; 50%=608.921us; 80%=008ms690.258us; 90%=014ms419.274us; 95%=020ms009.037us; 99%=045ms886.820us
Counter: CachedCompile
  Value: 1118
Counter: CreateCompileHandles
  Value: 4
Counter: CreateDataHandles
  Value: 25481
Counter: CreateXlaTensor
  Value: 18432587
Counter: DestroyDataHandles
  Value: 25465
Counter: DestroyXlaTensor
  Value: 18432565
Counter: DeviceDataCacheMiss
  Value: 4
Counter: MarkStep
  Value: 1122
Counter: ReleaseDataHandles
  Value: 25465
Counter: UncachedCompile
  Value: 4
Counter: XRTAllocateFromTensor_Empty
  Value: 27
Counter: XrtCompile_Empty
  Value: 128
Counter: XrtExecuteChained_Empty
  Value: 128
Counter: XrtExecute_Empty
  Value: 128
Counter: XrtRead_Empty
  Value: 128
Counter: XrtReleaseAllocationHandle_Empty
  Value: 128
Counter: XrtReleaseCompileHandle_Empty
  Value: 128
Counter: XrtSessionCount
  Value: 10
Counter: XrtSubTuple_Empty
  Value: 128
Counter: xla::_log_softmax
  Value: 1105
Counter: xla::_log_softmax_backward_data
  Value: 1105
Counter: xla::add
  Value: 2114970
Counter: xla::add_
  Value: 722663
Counter: xla::addmm
  Value: 708305
Counter: xla::as_strided
  Value: 41
Counter: xla::bernoulli_
  Value: 1105
Counter: xla::cat
  Value: 353600
Counter: xla::copy_
  Value: 41
Counter: xla::div_
  Value: 1105
Counter: xla::embedding
  Value: 1105
Counter: xla::embedding_dense_backward
  Value: 1105
Counter: xla::empty
  Value: 2251
Counter: xla::empty_strided
  Value: 41
Counter: xla::fill_
  Value: 1105
Counter: xla::index_select
  Value: 1105
Counter: xla::mm
  Value: 1415505
Counter: xla::mul
  Value: 3183505
Counter: xla::nll_loss_backward
  Value: 1105
Counter: xla::nll_loss_forward
  Value: 1105
Counter: xla::sigmoid_
  Value: 1060800
Counter: xla::sigmoid_backward
  Value: 1060800
Counter: xla::split
  Value: 353600
Counter: xla::stack
  Value: 4420
Counter: xla::sum
  Value: 708305
Counter: xla::t
  Value: 2832115
Counter: xla::tanh
  Value: 353600
Counter: xla::tanh_
  Value: 353600
Counter: xla::tanh_backward
  Value: 707200
Counter: xla::transpose
  Value: 4420
Counter: xla::unbind
  Value: 4420
Counter: xla::view
  Value: 713830
Counter: xla::zero_
  Value: 7728
Metric: XrtAllocateFromTensor
  TotalSamples: 63362
  Accumulator: 02m10s200ms767.219us
  Mean: 002ms075.807us
  StdDev: 778.413us
  Rate: 10.7113 / second
  Percentiles: 25%=001ms284.624us; 50%=002ms308.897us; 80%=003ms772.551us; 90%=003ms972.661us; 95%=003ms168.602us; 99%=004ms576.723us
Metric: XrtCompile
  TotalSamples: 98
  Accumulator: 10m54s320ms435.111us
  Mean: 06s064ms494.236us
  StdDev: 14s570ms744.908us
  Rate: 0.0164305 / second
  Percentiles: 25%=295.375us; 50%=021ms120.028us; 80%=266ms106.709us; 90%=35s640ms945.225us; 95%=39s602ms846.919us; 99%=39s935ms726.141us
Metric: XrtExecute
  TotalSamples: 21114
  Accumulator: 01h01m43s949ms833.777us
  Mean: 169ms349.511us
  StdDev: 095ms401.243us
  Rate: 3.82357 / second
  Percentiles: 25%=097ms848.351us; 50%=157ms248.064us; 80%=242ms428.398us; 90%=290ms943.523us; 95%=342ms207.421us; 99%=453ms905.782us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles: 
Metric: XrtReleaseAllocation
  TotalSamples: 209515
  Accumulator: 06s245ms288.928us
  Mean: 026.338us
  StdDev: 035.876us
  Rate: 44.1387 / second
  Percentiles: 25%=010.385us; 50%=015.109us; 80%=030.618us; 90%=061.331us; 95%=097.648us; 99%=190.847us
```

**Comment by btomtom5** [comment]
Oh I also turned it into a google colab notebook so that people can more easily experiment with it. [Here](https://colab.research.google.com/drive/1E_iRFsYJcVweLLPNPRHgbfqs2BInC3fk?usp=sharing)

**Comment by btomtom5** [comment]
Another question: What causes DeviceLockWaits? Are these the result of the core waiting for an xm.rendezvous to happen?

**Comment by dlibenzi** [comment]
I don't see anything wrong there.
Compilations stabilized, and the ExecuteTime 50% is about 150ms.
The IR graph, at 35K nodes is pretty big, but again, I see nothing wrong happening.

PyTorch/XLA overlaps device computation of step N, with pytorch python code of step N+1.
If the device step time is higher than the pytorch pytorch code execution time (usually is), you should see very little device lock waits.
If `DeviceLockWait` is non trivial, then the model has a fast device step time, compared to the python code execution time.

**Comment by jysohn23** [comment]
In that colab example it looks like you're using a colab VM that has 2 CPU cores. Instead can you try cloning one of these notebooks and running your code on that cloned notebook? https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/resnet18-training.ipynb

That should give you a better setup for PyTorch/XLA, especially if you're heavily input bound.

**Comment by btomtom5** [comment]
@jysohn23 Thanks for the suggestion. I'm going to try that right now, but what exactly is different about the notebook that you linked me that enables it to overcome the input bottleneck?

**Comment by jysohn23** [comment]
@btomtom5 Basically, we've set up the environment to be optimal for the pytorch/xla paradigm. It may have some more CPUs for input preprocessing to be able to keep the TPUs busy.

**Comment by btomtom5** [comment]
Right. I do see that you have fewer workers and the batch size is a multiple of 64, but maybe most importantly I see that the notebook you shared defaults to a high-ram tpu instance which appears to have 40 cpu cores instead of 2. I just tried switching the tpu instance on my code as is and I found that it ran 7/8x faster. That was some awesome advice! Thank you!

**Comment by jysohn23** [comment]
Great to hear. Closing the issue as all other metrics all look good. Feel free to reopen if needed.

**Comment by bjourne** [comment]
TensorFlow's TPU backend is much faster than PyTorch XLA on small Colab instances (2 CPUs). The time per epoch for a network comparable to the one in the gist with the same training data is 20 seconds.

The expectation is that PyTorch XLA and TensorFlow TPU should perform similarly on identical hardware. If that is not the case and PyTorch XLA cannot function well on small Colab instances it should be mentioned in the tutorials.

I don't know how to force Colab to give you large instances (40 CPUs) so I haven't been able to investigate that scenario fully. In the notebook btomtom5 created, the time per epoch for PyTorch with one GPU is 25 seconds. That is, almost the same as PyTorch XLA with 8 TPUs.

I'm unsure of if PyTorch XLA does automatic batch aggregation or not. TensorFlow TPU does, meaning that with batch size set to 32, batches of size 4 are sent to each TPU. In that scenario, PyTorch XLA on large instances takes 193 seconds per epoch.

**Comment by dlibenzi** [comment]
Please see this thread which shed some light on the temporary status of PyTorch vs TensorFlow in Cloud:

https://github.com/pytorch/xla/issues/1858

**Comment by bjourne** [comment]
I have now read that thread but I don't understand all of it. Is the gist of it that it is currently not possible to take full advantage of TPUs when using Google Colab, but that that will be fixed in the future?

**Comment by dlibenzi** [comment]
The `taking full advantage` depends on the model.
Many models are perfectly fine with Colab, but models which requires a lot of host computation resources might not.
In the next few months Google Cloud TPU will switch to the new architecture, where TF and PT will play an even game in that regard.

**Comment by bjourne** [comment]
Is my gist using a lot of host computation resources? I tried to make the training loop as efficient as possible and I can't see where it would be using a lot of resources.

**Comment by jysohn23** [comment]
The CPUs would be busy preprocesssing the input data, building the model graph in the backgroud, etc. The amount of CPU required to feed the TPU cores varies a lot by models (from a couple cores to ~100 CPU cores depending on dataset, augmentation pipeline, and model step time). 2 CPUs would easily be saturated.

**Comment by dlibenzi** [comment]
I see many time folks running many iterations over the same input dataset, and running the heavy input pipeline over and over.
In those cases it might be convenient to store the output of the input pipeline (GCS for example), and use the cooked version of it.
At that point your input pipeline will be just a read from storage, and a pytorch deserialization.

**Comment by bjourne** [comment]
Is there a guide somewhere on how to do that? I'd really like to use PyTorch for my project but it is not possible when it is so much slower than TF on the same hardware.

**Comment by dlibenzi** [comment]
I may have something by EOD today. Hold on ...

**Comment by dlibenzi** [comment]
https://github.com/pytorch/xla/pull/2275

**Comment by sipie800** [comment]
xla has many issues. It might be the only bad part of torch.
